File: Frontend/src/utils/qty.ts
// 与 Backend/app/core/qty.py 完全对齐的数量换算
// STACK = 64（一组）；SHULKER = 27 * 64 = 1728（一盒）
// 规则：n >= 1728 → "X盒"；n >= 64 → "X组"；否则 "X个"
// 换算系数取 round(n / X, 2) 后去尾零（等价 Python 的 :g 格式化）
const STACK = 64
const SHULKER = 27 * 64 // 1728

/**
 * 将数字格式化为「去尾零」字符串，等价 Python `f"{v:g}"`。
 * 例：2.0 → "2"；1.16 → "1.16"；1.56 → "1.56"；63 → "63"。
 */
function trimZero(v: number): string {
  return String(parseFloat(v.toFixed(2)))
}

export function formatQty(n: number): string {
  if (n >= SHULKER) return `${trimZero(n / SHULKER)}盒`
  if (n >= STACK) return `${trimZero(n / STACK)}组`
  return `${n}个`
}

File: McdrPlugin/pch_system/pch_system/qty.py
"""format_qty 数量换算纯函数（issue #18，与前后端三端对齐）。

权威源是后端 ``Backend/app/core/qty.py``，前端 ``Frontend/src/utils/qty.ts`` 已对齐；
本模块逐字照抄后端 ``STACK`` / ``SHULKER`` / ``format_qty``，保证三端字节级一致。
另提供显示层便利包装 ``format_qty_safe``（处理缺失/``"?"`` 兜底）。

Minecraft 物品单位换算（离线模式 MVP，无 NBT 容量差异）：
- SHULKER = 27 * 64 = 1728（一盒 = 一整个潜影盒 = 27 组）
- STACK   = 64（一组）

设计约束：
  * **纯函数**——仅依赖内置 ``int`` / ``round`` / f-string，不 import mcdreforged，
    可在无 MCDR 运行时的环境单测（与 ``text_layout.py`` / ``scanner.py`` 同范式）。
  * **显示层专用**——只换算展示串，不入库、不进 API 响应（DB ``need_qty`` 永存原始 int）。
"""

STACK: int = 64
SHULKER: int = 27 * STACK  # 1728


def format_qty(n: int) -> str:
    """将原始整数数量换算为「X盒 / X组 / X个」展示串。

    - n >= SHULKER → "X盒"（n/1728，保留两位去尾零，如 1.16盒）
    - n >= STACK   → "X组"（n/64）
    - 否则          → "X个"

    用 ``round(x, 2):g`` 去尾零：``2.0`` → ``"2"``，``1.50`` → ``"1.5"``。
    """
    if n >= SHULKER:
        return f"{round(n / SHULKER, 2):g}盒"
    if n >= STACK:
        return f"{round(n / STACK, 2):g}组"
    return f"{n}个"


def format_qty_safe(value) -> str:
    """显示层包装：``int`` 走 ``format_qty`` 换算；非 ``int``（如 ``"?"`` 缺失兜底）原样 ``str()``。

    用于 ``format_notification`` / 回执等用 ``payload.get(..., "?")`` 兜底的地方——
    ``format_qty("?")`` 会因 str 与 int 比较抛 ``TypeError``，故在此守护。
    ``bool`` 虽是 ``int`` 子类，但语义上不是数量，一并走 ``str()`` 兜底（避免 ``True``→``"True个"``）。
    """
    return format_qty(value) if isinstance(value, int) and not isinstance(value, bool) else str(value)
